modal/image.py

```diff
@@ -127,7 +127,7 @@ def __init__(
             raise InvalidError("Cannot run a build function with multiple base images!")
 
         for secret in secrets:
-            if not isinstance(secret, _Secret):
+            if not isinstance(secret, _Secret) and not isinstance(secret, Ref):
                 raise InvalidError(f"Secret {secret!r} must be a modal.Secret object")
 
         if context_mount is not None and not isinstance(context_mount, _Mount):
```

Comment:
Do we need to apply the same `Ref` check here in mounts?

Reply:
yes you're right. Let me add that in the follow-up PR with tests.

Reply:
(although I haven't seen anyone use a `Mount` as a ref, it's technically possible)

Reply:
I think we need to fix the type annotations on this library as well, where are `Ref`s allowed? How do we write code if no one knows what type things are

Reply:
Yeah we might want to define another type that's the `Union` of each object type and `Ref`? Not the cleanest solution. I think @erikbern had some ideas for having `Ref`s have better runtime types too.

Looks like we have this [cast](https://github.com/modal-labs/modal-client/blob/b2574e1ec16e554d823a6af8a982182cc659724c/modal/object.py#L180-L181) that casts `RemoteRef`s to the correct object type, but only for the type checker. Runtime types are unaffected (as we saw here).

Reply:
No, let's solve this at the foundational level – the correct solution is obviously that `Secret.from_name` returns a `Secret`

I don't know why it doesn't do that? Look at the signature in https://github.com/modal-labs/modal-client/blob/main/modal/object.py#L163

Reply:
It's because of the thing I mentioned above:

> Looks like we have this [cast](https://github.com/modal-labs/modal-client/blob/b2574e1ec16e554d823a6af8a982182cc659724c/modal/object.py#L180-L181) that casts RemoteRefs to the correct object type, but only for the type checker. Runtime types are unaffected (as we saw here).

It makes the type checker happy, but the runtime type is still `RemoteRef`.